### ISSUE TITLE
Delay in controlling BLE switch due to gattc queue full

### ISSUE THREAD

**Issue by narenarora** [esphome/issues#4205]
### The problem

I am using a esp32 board with esphome and switchmate config to control a switchmate toggle via BLE. Although it works, i always have a delay in the "turn on" or "turn off" commands due to continuous errors indicating gattc queue is full. Not sure how to fix this. Can one of you please help ? 

PS : I have no idea of the code for either project. I am just finding my way by searching others fixes.

### Which version of ESPHome has the issue?

2022.12.3

### What type of installation are you using?

Docker

### What platform are you using?

ESP32-IDF

### Board

generic esp32dev

### Component causing the issue

not sure

### Example YAML snippet

```yaml
substitutions:
  name: esp32-ble-proxy-1
packages:
  esphome.bluetooth-proxy: github://esphome/bluetooth-proxies/esp32-generic.yaml@main
esphome:
  name: ${name}
  name_add_mac_suffix: false

  includes:
    - switchmate.h

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

esp32_ble_tracker:
  scan_parameters:
    interval: 2000ms
    window: 1100ms
    active: true

bluetooth_proxy:
  active: true
  cache_services: true

api:
logger:
ota:
improv_serial:


# SwitchMate-ESPHome Config (requires switchmate.h)
ble_client:
  # You can find these through the Switchmate app
  - mac_address: <mac_addr>
    id: bedroom_light_esp

custom_component:
- lambda: |-
    return {
      new SwitchmateController("Lights (Controller)", false, 1000, bedroom_light_esp)
    };
  components:
  - id: lights_controller

switch:
- platform: custom
  lambda: |-
    return {
      ((SwitchmateController*)lights_controller)->state_switch
    };
  switches:
  - name: "Bedroom Light ESP"
    icon: "mdi:lightbulb-group"
# Switch to enable reboot of ESP32 device remotely
- platform: restart
  name: "Bedroom ESP32 Restart"

sensor:
  - platform: custom
    lambda: |-
      return {
        ((SwitchmateController*)lights_controller)->battery_sensor
      };
    sensors:
    - name: "Bedroom Lights ESP (Battery Level)"
      icon: 'mdi:battery'
      unit_of_measurement: '%'
      device_class: 'battery'
      accuracy_decimals: 2
      state_class: 'measurement'
      filters:
      - throttle: 60s
      - heartbeat: 60s
      #- debounce: 0.1s
      #filters:
      #- sliding_window_moving_average:
      #    window_size: 200
      #    send_every: 200
```

### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esp32-ble-proxy-1.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing esp32-ble-proxy-1 (board: esp32dev; framework: espidf; platform: platformio/espressif32 @ 5.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40402.0 (4.4.2) 
 - tool-cmake @ 3.16.9 
 - tool-ninja @ 1.10.2 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
Reading CMake configuration...
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- Improv @ 1.2.3
Compiling .pioenvs/esp32-ble-proxy-1/src/main.o
Linking .pioenvs/esp32-ble-proxy-1/firmware.elf
RAM:   [==        ]  16.7% (used 54632 bytes from 327680 bytes)
Flash: [=======   ]  65.7% (used 1205009 bytes from 1835008 bytes)
Building .pioenvs/esp32-ble-proxy-1/firmware.bin
Creating esp32 image...
Successfully created esp32 image.
esp32_create_combined_bin([".pioenvs/esp32-ble-proxy-1/firmware.bin"], [".pioenvs/esp32-ble-proxy-1/firmware.elf"])
Wrote 0x1379a0 bytes to file /config/.esphome/build/esp32-ble-proxy-1/.pioenvs/esp32-ble-proxy-1/firmware-factory.bin, ready to flash to offset 0x0
======================== [SUCCESS] Took 262.59 seconds ========================
INFO Successfully compiled program.
INFO Resolving IP address of esp32-ble-proxy-1.local
INFO  -> 192.168.0.55
INFO Uploading /config/.esphome/build/esp32-ble-proxy-1/.pioenvs/esp32-ble-proxy-1/firmware.bin (1210784 bytes)
Uploading: [============================================================] 100% Done...

INFO Waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from esp32-ble-proxy-1.local using esphome API
WARNING Can't connect to ESPHome API for esp32-ble-proxy-1.local: Error resolving IP address: [Errno -2] Name or service not known
INFO Trying to reconnect to esp32-ble-proxy-1.local in the background
INFO Successfully connected to esp32-ble-proxy-1.local
[21:31:59][I][app:102]: ESPHome version 2022.12.3 compiled on Feb 18 2023, 21:29:15
[21:31:59][I][app:104]: Project esphome.bluetooth-proxy version 1.0
[21:31:59][C][wifi:504]: WiFi:
[21:31:59][C][wifi:362]:   Local MAC: E0:5A:1B:75:BC:30
[21:31:59][C][wifi:363]:   SSID: [redacted]
[21:31:59][C][wifi:364]:   IP Address: 192.168.0.55
[21:31:59][C][wifi:366]:   BSSID: [redacted]
[21:31:59][C][wifi:367]:   Hostname: 'esp32-ble-proxy-1'
[21:31:59][C][wifi:369]:   Signal strength: -54 dB ▂▄▆█
[21:31:59][C][wifi:373]:   Channel: 3
[21:31:59][C][wifi:374]:   Subnet: 255.255.255.0
[21:31:59][C][wifi:375]:   Gateway: 192.168.0.1
[21:31:59][C][wifi:376]:   DNS1: 192.168.0.1
[21:31:59][C][wifi:377]:   DNS2: 0.0.0.0
[21:31:59][C][logger:293]: Logger:
[21:31:59][C][logger:294]:   Level: DEBUG
[21:31:59][C][logger:295]:   Log Baud Rate: 115200
[21:31:59][C][logger:296]:   Hardware UART: UART0
[21:31:59][C][bluetooth_proxy:065]: Bluetooth Proxy:
[21:31:59][C][bluetooth_proxy:066]:   Active: YES
[21:31:59][C][safe_mode.button:022]: Safe Mode Button 'Safe Mode Boot'
[21:31:59][C][safe_mode.button:022]:   Icon: 'mdi:restart-alert'
[21:31:59][C][restart:076]: Restart Switch 'Bedroom ESP32 Restart'
[21:31:59][C][restart:078]:   Icon: 'mdi:restart'
[21:31:59][C][restart:099]:   Restore Mode: restore defaults to OFF
[21:31:59][C][esp32_ble_tracker:870]: BLE Tracker:
[21:31:59][C][esp32_ble_tracker:871]:   Scan Duration: 300 s
[21:31:59][C][esp32_ble_tracker:872]:   Scan Interval: 2000.0 ms
[21:31:59][C][esp32_ble_tracker:873]:   Scan Window: 1100.0 ms
[21:31:59][C][esp32_ble_tracker:874]:   Scan Type: ACTIVE
[21:31:59][C][esp32_ble_tracker:875]:   Continuous Scanning: True
[21:31:59][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:31:59][C][ble_client:027]: BLE Client:
[21:31:59][C][ble_client:028]:   Address: FF:61:A0:8D:F7:9F
[21:31:59][C][mdns:103]: mDNS:
[21:31:59][C][mdns:104]:   Hostname: esp32-ble-proxy-1
[21:31:59][C][ota:093]: Over-The-Air Updates:
[21:31:59][C][ota:094]:   Address: esp32-ble-proxy-1.local:3232
[21:31:59][C][api:138]: API Server:
[21:31:59][C][api:139]:   Address: esp32-ble-proxy-1.local:6053
[21:31:59][C][api:143]:   Using noise encryption: NO
[21:31:59][C][improv_serial:032]: Improv Serial:
[21:32:00][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:01][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:02][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:03][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:04][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:05][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:06][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:07][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:08][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:09][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:10][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:11][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:12][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:13][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:14][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:15][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:16][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:17][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:18][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:19][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:20][D][esp32_ble_client:048]: [3] [FF:61:A0:8D:F7:9F] Found device
[21:32:20][D][esp32_ble_tracker:216]: Pausing scan to make connection...
[21:32:20][I][esp32_ble_client:064]: [3] [FF:61:A0:8D:F7:9F] 0x01 Attempting BLE connection
[21:32:20][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:20][D][esp-idf:000]: W (31509) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[21:32:21][I][switchmate:140]: [Lights (Controller)] Connected successfully!
[21:32:21][D][esp32_ble_tracker:327]: Starting scan...
[21:32:21][D][switch:056]: 'Bedroom Light ESP': Sending state OFF
[21:32:36][D][sensor:127]: 'Bedroom Lights ESP (Battery Level)': Sending state 27.00000 % with 2 decimals of accuracy
[21:32:38][D][esp-idf:000]: W (49012) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x8

[21:32:38][D][esp-idf:000]: W (49016) BT_APPL: gattc_conn_cb: if=3 st=0 id=3 rsn=0x8

[21:32:38][D][esp-idf:000]: W (49018) BT_APPL: gattc_conn_cb: if=4 st=0 id=4 rsn=0x8

[21:32:38][D][esp-idf:000]: W (49019) BT_APPL: gattc_conn_cb: if=5 st=0 id=5 rsn=0x8

[21:32:38][D][esp-idf:000]: W (49020) BT_APPL: gattc_conn_cb: if=6 st=0 id=6 rsn=0x8

[21:32:38][W][switchmate:211]: [Lights (Controller)] Error reading char at handle 14, status=133.
[21:32:38][W][switchmate:145]: [Lights (Controller)] Disconnected!
[21:32:38][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:39][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:40][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:41][D][esp32_ble_client:048]: [3] [FF:61:A0:8D:F7:9F] Found device
[21:32:41][D][esp32_ble_tracker:216]: Pausing scan to make connection...
[21:32:41][I][esp32_ble_client:064]: [3] [FF:61:A0:8D:F7:9F] 0x01 Attempting BLE connection
[21:32:41][W][switchmate:098]: [Lights (Controller)] Cannot poll, not connected.
[21:32:41][D][esp-idf:000]: W (52344) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[21:32:42][I][switchmate:140]: [Lights (Controller)] Connected successfully!
[21:32:42][D][esp32_ble_tracker:327]: Starting scan...
[21:33:35][D][esp-idf:000]: E (106102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:36][D][sensor:127]: 'Bedroom Lights ESP (Battery Level)': Sending state 27.00000 % with 2 decimals of accuracy
[21:33:37][D][esp-idf:000]: E (108102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:38][D][esp-idf:000]: E (109103) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:42][D][esp-idf:000]: E (113106) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:45][D][esp-idf:000]: E (116102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:46][D][esp-idf:000]: E (117106) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:49][D][esp-idf:000]: E (120102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:50][D][esp-idf:000]: E (121108) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:55][D][esp-idf:000]: E (126105) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:33:56][D][esp-idf:000]: E (127102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:04][D][esp-idf:000]: E (135102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:06][D][esp-idf:000]: E (137113) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:08][D][esp-idf:000]: E (139102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:09][D][esp-idf:000]: E (140102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:10][D][esp-idf:000]: E (141102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:11][D][esp-idf:000]: E (142102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:12][D][esp-idf:000]: E (143102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:16][D][esp-idf:000]: E (147103) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:20][D][esp-idf:000]: E (151103) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:20][D][esp-idf:000]: E (151108) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:24][D][esp-idf:000]: E (155102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:30][D][esp-idf:000]: E (161108) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:32][D][esp-idf:000]: E (163102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:33][D][esp-idf:000]: E (164102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:36][D][sensor:127]: 'Bedroom Lights ESP (Battery Level)': Sending state 27.00000 % with 2 decimals of accuracy
[21:34:36][D][esp-idf:000]: E (167104) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:38][D][esp-idf:000]: E (169102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:40][D][esp-idf:000]: E (171102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:41][D][esp-idf:000]: E (172102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:46][D][esp-idf:000]: E (177106) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:47][D][esp-idf:000]: E (178102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:50][D][esp-idf:000]: E (181103) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:51][D][esp-idf:000]: E (182102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:54][D][esp-idf:000]: E (185110) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:56][D][esp-idf:000]: E (187102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:34:59][D][esp-idf:000]: E (190109) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:03][D][esp-idf:000]: E (194102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:03][D][esp-idf:000]: E (194105) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:04][D][esp-idf:000]: E (195102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:06][D][esp-idf:000]: E (197102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:07][D][esp-idf:000]: E (198102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:08][D][esp-idf:000]: E (199102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:10][D][esp-idf:000]: E (201108) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:11][D][esp-idf:000]: E (202107) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:12][D][esp-idf:000]: E (203102) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:14][D][esp-idf:000]: E (205101) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.

[21:35:14][D][esp-idf:000]: E (205108) BT_APPL: bta_gattc_enqueue(), the gattc command queue is full.
```

**Comment by distante** [comment]
@narenarora did you found out a fix?

**Comment by narenarora** [comment]
> @narenarora did you found out a fix?

I just updated the ESPHome version and re-installed it on my ESP32. That fixed the issue for me.
I don't have the delay , or these messages in the logs anymore.
Although, the battery % seems to be stuck at 11% for me. That doesn't get updated. I have run through a couple batteries, and its still stuck at 11%.

Just FYI - I have updated the ESPHome version a couple times post this, and still no issues.
Currently I'm on 2023.11.3 and its still working okay for me.
